src/sinks/util/rusoto.rs

```diff
@@ -17,3 +26,60 @@ pub fn client(resolver: Resolver) -> crate::Result<Client> {
 
     Ok(HttpClient::from_connector(https))
 }
+
+pub trait RusotoNewClient {
+    fn new_client<P>(client: Client, credentials_provider: P, region: Region) -> Self
+    where
+        P: ProvideAwsCredentials + Send + Sync + 'static,
+        P::Future: Send;
+}
+
+pub fn create_client<T: RusotoNewClient>(
```

Comment:
I find this trait a bit overkill for this imo. I think it'd be better to let each sink construct their concrete type via `T::new_with` within the sink, then have one fn to return the client and one fn to return a credentials provider. Then from there we can do something like

```rust
pub enum AwsCredentialsProvider {
	Default(DefaultCredentialsProvider),
	AssumeRole(...),
	Static(...),
}

fn client(...) -> crate::Result<Client> { ... }
fn credentials(...) -> crate::Result<AwsCredentialsProvider> { ... }
```

Imo this would be a bit easier to grok in a per sink sense and would avoid a set of macros and a trait.

Reply:
It took a little more work this way, but the approach makes sense and is more composable. I'll see about adding the static provider to avoid that block in the S3 sink.